backend/frontend/coursecreation/webapi.js

```diff
@@ -0,0 +1,224 @@
+    data: data,
+    dataType: "json"
+  }));
+}
+
+function deleteNode(courseId, nodeId: string){
+  let endpoint = URL + `/${courseId}/node/delete/${nodeId}/`;
+  return Promise.resolve($.ajax(endpoint, {
+    method: "POST",
+    dataType: "json"
+  }));
+}
+
+function addRoot(courseId, rootId){
+  rootId = rootId.toString();
+
+  let endpoint = URL + `/${courseId}/root/set/${rootId}/`;
+  return Promise.resolve($.ajax(endpoint, {
+    method: "POST"
+  }));
+}
+
+function initializeCourse(courseName){
+  if(courseName === ""){
+    courseName = "CS130";
+  }
+
+  let endpoint = URL + `/0/course/add/`;
+  return Promise.resolve($.ajax(endpoint, {
+    method: "POST",
+    data: {name: courseName},
+    dataType: "json"
+  }));
+}
+
+function setPiazza(courseId, piazzaId){
+  let data = {
+    "piazza_cid": piazzaId
+  };
+
+  let endpoint = URL + `/${courseId}/course/setpiazza/`;
+  return Promise.resolve($.ajax(endpoint, {
+    method: "POST",
+    data,
+    dataType: "json"
+  }));
+}
+
+//===================================
+
+function handleError(errorStr){
+  return (jqXHR, textStatus, errorThrown) => {
+    console.error(errorStr, jqXHR, textStatus, errorThrown);
+    throw errorThrown;
+  }
+}
+
+
+//=================================
+
+function flatten(course: Object){
+  let nodes = [];
+
+  let courseNode = {};
+  courseNode.renderer = "Timeline";
+  courseNode.contents = "";
+  courseNode.children = pairsToObject(course.weeks.map((week, i)=> weekToNode(week, i)));
+
+  return bfs(courseNode);
+}
+
+function bfs(root){
+  let workingQueue = [root];
+  let returnQueue = [];
+
+  while(!(workingQueue.length === 0)){
+    let current = workingQueue.shift();
+    returnQueue.push(current);
+
+    for(let tag in current.children){
+      workingQueue.push(current.children[tag]);
+    }
+  }
+
+  return returnQueue;
+}
+
+function weekToNode(week, i){
+  let weekNode = {
+    contents: "",
+    renderer: "Week",
+    children: {
+      announcements : {
+        contents: JSON.stringify(week.dateRangeInput),
+        renderer: "Announcements",
+        children: {}
+      },
+      assignments : assignmentsToNode(week.assignments),
+      topics : topicsToNode(week.topics)
+    }
+  }
+  return [`Week ${i+1}`, weekNode];
+}
+
+function assignmentsToNode(assignments){
+  return {
+    contents: "Assignments\n==",
+    renderer: "ModalList",
+    children: pairsToObject(assignments.map((assign)=> [assign.title, {
+      contents: assign.markdown,
+      renderer: "Assignment",
+      children: {}
+    }]))
+  };
+}
+
+function topicsToNode(topics){
+  return {
+    contents: "",
+    renderer: "ModalList",
+    children: pairsToObject(topics.map(topicToNode))
+  };
+}
+
+function topicToNode(topic){
+  return [topic.title.toLowerCase(), {
+    contents: `${topic.title}\n==`,
+    renderer: "Topic",
+    children: {
+      resources: resourcesToNode(topic.resources)
+    }
+  }];
+}
+
+function resourcesToNode(resources){
```

Comment:
If you create a topic with a name, but leave its resources blank, this causes issues (you can try editing a nonexistent resource, which is no good). This should be resolved.